### ISSUE TITLE
Error with calcdiversity output

### ISSUE THREAD

**Issue by bmillerlab** [mennodejong1986/SambaR#24]
Hello, I've been running sambar using 

>source("https://github.com/mennodejong1986/SambaR/raw/master/SAMBAR_v1.08.txt")

I am copying the mypackageslist.txt output file as an attached file, though I don't see any problems there that seem to have anything to do with the error I'm getting.

My dataset imports without any issues and I can run the filterdata and findstructure commands without problems:

>importdata(inputprefix="all", do_citations=FALSE, sumstatsfile=FALSE , depthfile=FALSE , samplefile="popmap.txt", colourvector=colvector1, pop_order=poporder)  

>filterdata(indmiss=0.9,snpmiss=0.1,min_mac=2,dohefilter=TRUE, maxprop_hefilter=0.5, min_spacing=500,nchroms=NULL,TsTvfilter=NULL, F_correct_maf=FALSE)
>findstructure(Kmax=10,add_legend=TRUE,legend_pos="right",legend_cex=1, quickrun=TRUE, doBAPT=FALSE)

Problems completing analysis begin with the calcdiversity command: 

>calcdiversity(nrsites=NULL,legend_cex=0.5, dohwe=FALSE) 
Because the flag do_SFS is set to FALSE (default), SambaR will not generate SFS-vectors.
REMINDER: if you receive an error stating 'cannot open file' (without the addition 'No such file or directory') this is likely because the file is opened in another file viewer.
If so, close the relevant file in the file viewer and try again.
Hardy Weinberg calculations...
If you receive an error after this line, set the flag dohwe to FALSE.
Plotting multilocus heterozygosity...
Column 'MLH' not present in inds dataframe. Setting it equal to inds$hetero_all.
Error in boxplot.default(split(mf[[response]], mf[-response], drop = drop,  : 
  invalid first argument
Called from: boxplot.default(split(mf[[response]], mf[-response], drop = drop, 
    sep = sep, lex.order = lex.order), xlab = xlab, ylab = ylab, 
    add = add, ann = ann, horizontal = horizontal, ...)
Browse[1]> 

When I look at the output from Browse[1]>  it highlights this line:

if (0L == (n <- length(groups))) 
stop("invalid first argument")

I am not sure what I need to do to fix this problem and if I try and continue running the additional commands for sambar I keep getting errors (though parts of the analysis will complete).  Is there a problem that I can fix?  

Thank you for any help you can offer.

[mypackageslist.txt](https://github.com/mennodejong1986/SambaR/files/9577729/mypackageslist.txt)

**Comment by mennodejong1986** [comment]
Hi, so far I did not manage to reproduce the error. However, I did add a few more lines of screen output which might help to trace the cause of the problem. Could you rerun the calcdiversity() function (after sourcing SambaR again) and let me know the screen output?

**Comment by bmillerlab** [comment]
Hi, Thanks for trying to help me.

It looks the same to me, but here is the output:

> calcdiversity(nrsites=NULL,legend_cex=0.5) 
Closing open devices...
Because the flag do_SFS is set to FALSE (default), SambaR will not generate SFS-vectors.
REMINDER: if you receive an error stating 'cannot open file' (without the addition 'No such file or directory') this is likely because the file is opened in another file viewer.
If so, close the relevant file in the file viewer and try again.
Hardy Weinberg calculations...
If you receive an error after this line, set the flag dohwe to FALSE.
Plotting multilocus heterozygosity...
Column 'MLH' not present in inds dataframe. Setting it equal to inds$hetero_all.
Error in boxplot.default(split(mf[[response]], mf[-response], drop = drop,  : 
  invalid first argument
Called from: boxplot.default(split(mf[[response]], mf[-response], drop = drop, 
    sep = sep, lex.order = lex.order), xlab = xlab, ylab = ylab, 
    add = add, ann = ann, horizontal = horizontal, ...)
Browse[1]> 

AND then in the file "boxplot.default" that opens: There is a green arrow at line 15 pointing to the line that is highlighted in yellow (stop.....):

if (0L == (n <- length(groups))) 
    stop("invalid first argument")


I don't know code in R well enough to understand what "groups" are referred to, but I did go back and check this time and the "MLH_boxplot.eps" file is created but is an empty file.

**Comment by bmillerlab** [comment]
Hi, Ok I went back and looked at the loaded packages and realized that inbreedR was not installed with sambar (not sure why as I don't get an error message about that).  So I installed the package and now the screen reads:

> calcdiversity(nrsites=NULL,legend_cex=0.5) 
Closing open devices...
Because the flag do_SFS is set to FALSE (default), SambaR will not generate SFS-vectors.
REMINDER: if you receive an error stating 'cannot open file' (without the addition 'No such file or directory') this is likely because the file is opened in another file viewer.
If so, close the relevant file in the file viewer and try again.
Hardy Weinberg calculations...
If you receive an error after this line, set the flag dohwe to FALSE.
Calculating (standardized) multilocus heterozygosity (sMLH) using inbreedR...
Plotting standardized multilocus heterozygosity...
Error in boxplot.default(split(mf[[response]], mf[-response], drop = drop,  : 
  invalid first argument
Called from: boxplot.default(split(mf[[response]], mf[-response], drop = drop, 
    sep = sep, lex.order = lex.order), xlab = xlab, ylab = ylab, 
    add = add, ann = ann, horizontal = horizontal, ...)

The error still points to the same lines in "boxplot.default."

**Comment by mennodejong1986** [comment]
This suggests the problem is somehow related to the MLHboxplot function (which is invoked by the calcdiversity function), more specifically when trying to output the plot in eps format. What happens when you run the MLHboxplot function by itself and without exporting? 

To find out, run:
MLHboxplot(export=NULL,sMLH=FALSE,plotname="MLH_boxplot",ylabel="Multi locus heterozygosity")

And can you export in another format (for example 'pdf'):
MLHboxplot(export="pdf",sMLH=FALSE,plotname="MLH_boxplot",ylabel="Multi locus heterozygosity")

**Comment by mennodejong1986** [comment]
The MLHboxplot function plots column inds$MLH (or, if absent, inds$hetero_all) against the column inds$poporder. What happens when you plot this yourself:

boxplot(inds$MLH~inds$poporder)
stripchart(MLH~poporder,data=inds,vertical=TRUE,add=TRUE)

Do both columns exist, and do you use anything unusual? The inds$MLH column should be numeric (containing floats) while the poporder column should be of the class factor.

**Comment by mennodejong1986** [comment]
For some reason, the commands are not displayed properly in the previous comment. You can also have a look within the sambar script at the MLHboxplot function to find the correct command (simple boxplot function).

**Comment by bmillerlab** [comment]
Hello, I tried all the commands suggested and here is the output for the errors after each one:

>MLHboxplot(export=NULL,sMLH=FALSE,plotname="MLH_boxplot",ylabel="Multi locus heterozygosity")
Error in boxplot.default(split(mf[[response]], mf[-response], drop = drop,  : 
  invalid first argument
Called from: boxplot.default(split(mf[[response]], mf[-response], drop = drop, 
    sep = sep, lex.order = lex.order), xlab = xlab, ylab = ylab, 
    add = add, ann = ann, horizontal = horizontal, ...)

> MLHboxplot(export="pdf",sMLH=FALSE,plotname="MLH_boxplot",ylabel="Multi locus heterozygosity")
Error during wrapup: invalid first argument
Error: no more error handlers available (recursive errors?); invoking 'abort' restart
Browse[1]> 

> boxplot(inds$MLHinds$poporder)
Error during wrapup: need finite 'ylim' values
Error: no more error handlers available (recursive errors?); invoking 'abort' restart
Browse[1]> 

> stripchart(MLHpoporder,data=inds,vertical=TRUE,add=TRUE)
Error during wrapup: object 'MLHpoporder' not found
Error: no more error handlers available (recursive errors?); invoking 'abort' restart
Browse[1]> 

Each time the Browse[1]>  opened the same file (boxplot.default) and pointed to the same line as above. 


So I looked in the original script.   There seem to be a lot of specific instructions in the script about what to do if using Windows that don't apply to me, since I'm on a mac.  However, looking through the commands made me wonder if this problem is because I have put in a specific order for my population now:

importdata(inputprefix="all", do_citations=FALSE, sumstatsfile=FALSE , depthfile=FALSE , samplefile="popmap.txt", geofile="geofile_sample_names.txt", colourvector=colvector1, pop_order=poporder) 

When I ran this type analysis before on another population (different data), I did not run into this problem.  For this new dataset I input geographical data and wanted the population in a specific order in the output files.  I will rerun the analysis without this option and see if it works that way.

**Comment by bmillerlab** [comment]
Hello, Well it does not seem to have to do with population data.  I went back and tried rerunning my other dataset and now it has the same problem too.  I am not good with code but I did go back and look at the code:

		cat("Calculating (standardized) multilocus heterozygosity (sMLH) using inbreedR...",sep="\n")
		calcMLH()
		cat("Plotting standardized multilocus heterozygosity...",sep="\n")
		MLHboxplot(export="eps",sMLH=FALSE,plotname="MLH_boxplot",ylabel="Multi locus heterozygosity")
		MLHboxplot(export="pdf",sMLH=FALSE,plotname="MLH_boxplot",ylabel="Multi locus heterozygosity")
		MLHboxplot(export="png",sMLH=FALSE,plotname="MLH_boxplot",ylabel="Multi locus heterozygosity")

And the one thing I could figure out was that I needed to have the calcMLH() output to get the next steps to work and it did not seem it existed.  Though it did not say that when I tried running your suggested command originally:
MLHboxplot(export=NULL,sMLH=FALSE,plotname="MLH_boxplot",ylabel="Multi locus heterozygosity")
(see output in prior message above).

So, this time I restarted R (didn't save workspace) and imported the dataset and filtered it but stopped there and tried running it again without having run calcdiversity at all and got:
> MLHboxplot(export=NULL,sMLH=TRUE,plotname="MLH_boxplot",ylabel="Multi locus heterozygosity")
Column 'MLH' not present in inds dataframe. Setting it equal to inds$hetero_all.
Column 'sMLH' not present in inds dataframe. First run the calcMLH function.

So, I  went ahead and reran it this way:
>calcMLH(popnames=mysambar$populations,phenovector=NULL)

>MLHboxplot(export=NULL,sMLH=FALSE,plotname="MLH_boxplot",ylabel="Multi locus heterozygosity")

This worked, but did not output a plot file, just an image so I tried (outputs file to sambar3 directory)
>MLHboxplot(export="pdf",sMLH=FALSE,plotname="MLH_boxplot",ylabel="Multi locus heterozygosity")


So for some reason when I'm running the regular calcdiversity() command it is not including this step in the output.  However, there isn't anything wrong with my dataframe or anything, since I can actually run the commands correctly.

Clearly I can work around this issue now, but I don't understand why this is happening.

**Comment by mennodejong1986** [comment]
Great trouble shooting and thank you for this useful information.
Apparantly some of the earlier steps of the calcdiversity function cause a small problem   

I just uploaded a new script (still called Sambar_v1.08.txt) in which I changed the order of commands in the calcdiversity function. It now first generates the heterozygosity boxplot functions before running other type of analyses.  
 
I also added two more switches which can be controlled with the do_he flag and do_triplot flags. These flags are by default set to TRUE, but if the error still persists (but now likely for another plot, given that the heterozygosity boxplots will be generated first), you could try rerunning the calcdiversity function with the flag 'do_triplot' set to FALSE. 

As a side note:
If the output from calcMLH does not exist (for example because the package inbreedR has not been installed), the MLHboxplot function uses the column inds$hetero_all (instead of the column inds$MLH) as input. The values of inds$hetero_all have been generated with a SambaR function, whereas inds$MLH has been generated with a function of InbreedR. These two functions give identical results because they calculate the same thing. 

In other words, the function MLHboxplot can be run even without first running the calcMLH function.  If so, the following line is printed on the screen (see also your very first comment): 
Column 'MLH' not present in inds dataframe. Setting it equal to inds$hetero_all.
